ComponentKit/Utilities/CKComponentTrigger.h

```diff
@@ -0,0 +1,169 @@
+/*
+ *  Copyright (c) 2014-present, Facebook, Inc.
+ *  All rights reserved.
+ *
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#import <Foundation/Foundation.h>
+
+#import <memory>
+
+#import <ComponentKit/CKComponentScope.h>
+#import <ComponentKit/CKComponentAction.h>
+
+/**
+ A component trigger is the opposite of an action. It is used by a parent component to trigger some side effect in a
+ child component.
+ 
+ Similar to an action, it is created in a parent scope, and is threaded down to a child component. Unlike an action
+ though, triggers activate **downward** from a parent component onto a child. Actions are triggered by a child, and
+ cause side-effects in parent components.
+ 
+ Triggers are an alternative to manually exposing methods on parent components that invoke on child components. Instead
+ triggers can be passed through many layers of components.
+ 
+ Triggers use the same pattern as component actions to allow parameters to be passed to targets. The first argument to
+ a trigger selector will always be the sender of the trigger invocation, the next arguments types are determined by
+ the variadic template types. This allows passing of data down to the trigger invocation method.
+ 
+ Triggers passed into a component **must** be resolved once, and only once. Resolution assigns a target and selector
```

Comment:
> Triggers passed into a component **must** be resolved once, and only once.

I'm a little concerned by this restriction. Any descendant of the parent can resolve a trigger. The runtime assertion helps but at the end of the day I feel like this may be a common issue. Worst case triggers for particular components may just stop working if another descendant resolves it instead. In that case which one should actually "win"?